packagedb/find_source_repo.py

```diff
@@ -0,0 +1,454 @@
+        )
+        if source_urls:
+            yield from source_urls
+
+    if package.type == "maven":
+        yield from get_source_urls_from_package_data_and_resources(
+            package=get_merged_ancestor_package_from_maven_package(package=package)
+        )
+
+
+def get_source_urls_from_package_data_and_resources(package: Package) -> List[str]:
+    """
+    Return the URL of the source repository of a package
+    or None if not found
+    """
+    if not package:
+        return []
+    source_urls = list(get_urls_from_package_data(package))
+    if source_urls:
+        return source_urls
+    source_urls = list(get_urls_from_package_resources(package))
+    if source_urls:
+        return source_urls
+    return []
+
+
+def convert_repo_urls_to_purls(source_urls):
+    """
+    Convert a source URL to a purl
+    """
+    url_hints = [
+        "github",
+        "gitlab",
+        "bitbucket",
+    ]
+    if not source_urls:
+        return
+    for source_url in source_urls:
+        # git@github.com+https://github.com/graphql-java/java-dataloader.git
+        if source_url.startswith("git@github.com+"):
+            _, _, source_url = source_url.partition("+")
+        # https+//github.com/graphql-java-kickstart/graphql-java-servlet.git
+        if source_url.startswith("https+//"):
+            # convert https+// to https://
+            source_url = source_url.replace("https+//", "https://")
+        if (
+            source_url.startswith("git+https://") or source_url.startswith("git://")
+        ) and "@" in source_url:
+            # remove the commit from the end of the URL
+            source_url, _, _ = source_url.rpartition("@")
+        # remove .git from the end of the URL
+        if source_url.endswith(".git"):
+            source_url, _, _ = source_url.rpartition(".git")
+        if source_url.startswith("git://"):
+            # remove git:// from the beginning of the URL
+            _, _, source_url = source_url.partition("git://")
+            if ":" in source_url:
+                # convert : to /
+                source_url = source_url.replace(":", "/")
+            source_url = f"https://{source_url}"
+        urlparse_result = urlparse(source_url)
+        path_segments = urlparse_result.path.split("/")
+        if not len(path_segments) > 2:
+            continue
+        namespace = path_segments[1]
+        name = path_segments[2]
+        if not name:
+            continue
+        for url_hint in url_hints:
+            if url_hint in urlparse_result.netloc:
+                yield PackageURL(
+                    type=url_hint,
+                    namespace=namespace,
+                    name=name,
+                )
+
+
+def get_urls_from_package_resources(package):
+    """
+    Return the URL of the source repository of a package
+    """
+    urls = []
+    for resource in package.resources.filter(is_key_file=True):
+        urls = [url["url"] for url in resource.urls]
```

Comment:
Bug here: you are continuously overwriting `urls` on each iteration, only the urls from the last resource that had urls will be saved to `urls`. Do you mean to append or extend `urls` on each iteration?

Reply:
Good catch, I think I missed the indentation at line 315, added the tests for same as well